File: Introduction_to_numerical_programming_using_Python_and_CPP_Beu/Ch10/Python/P10-qTrapz.py
# Evaluates an integral using the trapezoidal rule
from math import *

def Func(x): return (x*x*x) * exp(-x)

#============================================================================
def qTrapz(Func, a, b, n):
#----------------------------------------------------------------------------
#  Integrates function Func on interval [a,b] using the trapezoidal rule
#  with n integration points
#----------------------------------------------------------------------------
   h = (b-a)/(n-1)
   s = 0.5*(Func(a) + Func(b))
   for i in range(1,n-1): s += Func(a+i*h)

   return h*s

# main

a = 0e0; b = 1e0; n = 100
print("I = ",qTrapz(Func,a,b,n))

File: Introduction_to_numerical_programming_using_Python_and_CPP_Beu/Ch10/C/P10-qTrapz.cpp
// Evaluates an integral using the trapezoidal rule
#include <stdio.h>
#include <math.h>

double Func(double x) { return (x*x*x) * exp(-x); }

//===========================================================================
double qTrapz(double Func(double), double a, double b, int n)
//---------------------------------------------------------------------------
// Integrates function Func on interval [a,b] using the trapezoidal rule
// with n integration points
//---------------------------------------------------------------------------
{
   double h, s;
   int i;

   h = (b-a)/(n-1);
   s = 0.5*(Func(a) + Func(b));
   for (i=1; i<=(n-2); i++) s += Func(a+i*h);
   
   return h*s;
}

int main()
{
   double a, b;
   int n;
   
   a = 0e0; b = 1e0; n = 100;
   printf("I = %f\n",qTrapz(Func,a,b,n));
}
